File: LeetCode/2257. Count Unguarded Cells in the Grid/solution.cpp
class Solution {
public:
    int countUnguarded(int n, int m, vector<vector<int>>& guards, vector<vector<int>>& walls) {
        vector<vector<bool>> G, W, f;
        G.resize(n, vector<bool>(m));
        W.resize(n, vector<bool>(m));
        f.resize(n, vector<bool>(m));
        for (auto &g : guards) G[g[0]][g[1]] = true;
        for (auto &w : walls) W[w[0]][w[1]] = true;

        #define GAO { \
            if (G[i][j]) now = true; \
            else if (W[i][j]) now = false; \
            if (now) f[i][j] = true; \
        }

        for (int i = 0; i < n; i++) {
            bool now = false;
            for (int j = 0; j < m; j++) GAO
            now = false;
            for (int j = m - 1; j >= 0; j--) GAO
        }
        for (int j = 0; j < m; j++) {
            bool now = false;
            for (int i = 0; i < n; i++) GAO
            now = false;
            for (int i = n - 1; i >= 0; i--) GAO
        }

        int ans = 0;
        for (int i = 0; i < n; i++) for (int j = 0; j < m; j++) if (!G[i][j] && !W[i][j] && !f[i][j]) ans++;
        return ans;
    }
};


// 作者：tsreaper
// 链接：https://leetcode.cn/problems/count-unguarded-cells-in-the-grid/solution/by-tsreaper-znyl/
// 来源：力扣（LeetCode）
// 著作权归作者所有。商业转载请联系作者获得授权，非商业转载请注明出处。

File: LeetCode/2257. Count Unguarded Cells in the Grid/solution.py
class Solution:
    def countUnguarded(self, n: int, m: int, guards: List[List[int]], walls: List[List[int]]) -> int:
        grid = [['.']*m for _ in range(n)]
        for r, c in guards:
            grid[r][c] = 'G'
        for r, c in walls:
            grid[r][c] = 'W'

        def operate(r,c,flag):
            if grid[r][c] == 'W':
                flag = False
            elif grid[r][c] == 'G':
                flag = True
            elif flag:
                grid[r][c] = '#'
            return flag

        for r in range(n):
            flag = False
            for c in range(m):
                flag = operate(r,c,flag)
            flag = False
            for c in range(m)[::-1]:
                flag = operate(r,c,flag)
        for c in range(m):
            flag = False
            for r in range(n):
                flag = operate(r,c,flag)
            flag = False
            for r in range(n)[::-1]:
                flag = operate(r,c,flag)
        return sum(ch == '.' for row in grid for ch in row)

                
                
                
                
                
                



class Solution:
    def countUnguarded(self, n: int, m: int, guards: List[List[int]], walls: List[List[int]]) -> int:
        guards = set([(r,c) for r, c in guards])
        walls = set([(r,c) for r, c in walls])
        visited = set()
        visible = False
        # print(guards,walls)

        def doit(i,j):
            nonlocal visible
            if (i,j) in guards or (i,j) in walls:
                # print(i,j,"gw")
                visited.add((i,j))
            if (i,j) in guards:
                visible = True
            if (i,j) in walls:
                visible = False
            if visible:
                # print(i,j,"visible")
                visited.add((i,j))

        for i in range(n):
            visible = False
            for j in range(m):
                doit(i,j)
            visible = False
            for j in range(m)[::-1]:
                doit(i,j)            
            
            
        for j in range(m):
            visible = False
            for i in range(n):
                doit(i,j)
            visible = False
            for i in range(n)[::-1]:
                doit(i,j)

        return m*n-len(visited)